### ISSUE TITLE
allow specifying types instead of field names in `field_validator`s

### ISSUE THREAD

**Issue by DetachHead** [pydantic/pydantic#6611]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

the most common use case i have for `field_validator` is to validate every field with a certain type (usually `date` and `datetime`), but it's annoying to have to specify every date field twice:
```py
from datetime import date

from pydantic import field_validator
from pydantic.dataclasses import dataclass


@dataclass
class Foo:
    start_date: date
    end_date: date
    some_other_date: date

    @field_validator("start_date", "end_date", "some_other_date")
    @staticmethod
    def _parse_date(value: str) -> date:
        """this api stores dates in a wacky format so we need a custom date parser"""
```

it would be cool if you could specify the type in the decorator like this instead:
```py
@field_validator(date)
@staticmethod
def _parse_date(value: str) -> date:
    ...
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

Selected Assignee: @samuelcolvin

**Comment by adriangb** [comment]
Does this do what you want?

```python
from datetime import date
from typing import Annotated

from pydantic import BaseModel, BeforeValidator

MyDate = Annotated[date, BeforeValidator(lambda x: x.replace('2021', '2023'))]


class Foo(BaseModel):
    start_date: MyDate


print(Foo.model_validate_json('{"start_date": "2021-01-01"}'))
#> start_date=datetime.date(2023, 1, 1)
```

**Comment by DetachHead** [comment]
thanks, that's better than what i've been doing. though it would still be nice to define the validator on the class for all `date`s without having to define a new type for it.

**Comment by adriangb** [comment]
That would be #6535

**Comment by KotlinIsland** [comment]
@adriangb That thread seems really hard to follow. Is this functionality currently available in Pydantic?

I have a subtype of `date` and I want to add a validator on it to support `fromisoformat` like Pydantic does with `date` (#8122)

```py
from datetime import date
from pydantic import BaseModel
class MyDate(date): ...

class Model(BaseModel):
    d: MyDate
Model(d="2022-02-01")  # Unable to generate pydantic-core schema

# Hacky moment
MyDate = Annotated[MyDate, BeforeValidator(MyDate.fromisoformat)]
```

Any idea if this is possible?
